In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('telco_customer_churn.csv')

In [ ]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


### Welcome to Day 9 Homework. Please answer the questions below!

1. Buanglah baris dengan `tenure` samadengan 0 (jadi, data yang akan kita analisa di PR ini adalah baris-baris dengan `tenure` TIDAK samadengan 0). Ada berapa baris yang sekarang tersedia di data kita?

2. Bagi customer yang ***TIDAK MEMILIKI PARTNER*** dan ***TIDAK MEMILIKI DEPENDENTS***, berapa ***rata-rata*** dari `tenure` mereka?

3. Bagi customer yang ***MEMILIKI PARTNER*** dan ***TIDAK MEMILIKI DEPENDENTS***, berapa maksimal `MonthlyCharges` yang tercata?

4. Berapa banyak customer yang ***TIDAK MEMILIKI PARTNER*** namun ***MEMILIKI DEPENDENTS***?

5. Berapa rata-rata ***TOTAL CHARGES*** dari customer ***NON-SENIOR CITIZEN*** (Senior Citizen = 0) yang memiliki ***PHONE SERVICE*** dan memiliki ***INTERNET SERVICE FIBER OPTIC***?

#### Silakan menuliskan kode untuk menjawab 5 pertanyaan di atas...

Jumlah baris pada dataset diamati

In [ ]:
df.shape[0]

7043

Identifikasi nilai tenur diatas 0 untuk memastikan jumlah yang berlangganan

In [ ]:
df_tenure = df[df['tenure']>0]
df_tenure.shape[0]

7032

Sehingga diketahui bahwa yang berlangganan sebanyak 7032 namun yang tidak berlangganan sebanyak 11

- Selanjutnya mencari Average atau rata-rata `tenure` untuk customer yang ***TIDAK MEMILIKI PARTNER*** dan ***TIDAK MEMILIKI DEPENDENTS***

In [ ]:
df_tenure.groupby(['Partner','Dependents']).agg({'tenure':'mean',
                                                 #'MonthlyCharges':'max',
                                                 }).reset_index()

Partner Dependents     tenure
0      No         No  23.303354
1      No        Yes  23.986072
2     Yes         No  42.709014
3     Yes        Yes  41.578161

- Selanjutnya ditambahkan dengan bagi customer yang ***MEMILIKI PARTNER*** dan ***TIDAK MEMILIKI DEPENDENTS*** melalui Maksimal

In [ ]:
df_tenure.groupby(['Partner','Dependents']).agg({'tenure':'mean',
                                                 'MonthlyCharges':'max',
                                                 }).reset_index()

Partner Dependents     tenure  MonthlyCharges
0      No         No  23.303354          118.65
1      No        Yes  23.986072          116.15
2     Yes         No  42.709014          118.60
3     Yes        Yes  41.578161          118.75

*Sehingga* nilai rata-rata `tenure` untuk customer yang tidak memiliki partner dan tidak memiliki dependents adalah 23.303. Nilai Maksimum `MonthlyCharges` untuk customer yang memiliki partner namun tidak memiliki dependents adalah 118.60

Selanjutnya bagi customer yang **TIDAK MEMILIKI PARTNER** dan **MEMILIKI DEPENDENTS**

In [ ]:
# Menggunakan fungsi Count untuk menghitung jumlah data yang kita inginkan
df_tenure.groupby(['Partner','Dependents'])['customerID'].count()

Partner  Dependents
No       No            3280
         Yes            359
Yes      No            1653
         Yes           1740
Name: customerID, dtype: int64

Terlihat sebanyak 359 Customer yang **tidak memiliki Partner namun memiliki dependent**

Selanjutnya menghitung nilai rata - rata **TOTAL CHARGES** dari customer **NON-SENIOR CITIZEN (Senior Citizen = 0)** yang memiliki **PHONE SERVICE** dan memiliki **INTERNET SERVICE FIBER OPTIC**?

In [ ]:
# Jika format total charges tidak diubah dalam bentuk Float maka akan terjadi eror
# Hal ini karena Bentuk format Total Charges adalah Object sehingga harus dijadikan numerik
# Contoh error Langsung menggunakan fungsi bentuk Object

# df_tenure.groupby(['SeniorCitizen','PhoneService','InternetService'])['TotalCharges'].mean()

# Lalu diubah terlebih dahulu ke Float
df_tenure['TotalCharges'] = df_tenure['TotalCharges'].astype(float)
df_tenure.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [ ]:
# Lalu kembali menggunakan fungsi awal namun dengan data yang sudah diiubah ke float, 
df_tenure.groupby(['SeniorCitizen','PhoneService','InternetService'])['TotalCharges'].mean()

SeniorCitizen  PhoneService  InternetService
0              No            DSL                1508.513021
               Yes           DSL                2344.751771
                             Fiber optic        3226.492737
                             No                  655.759775
1              No            DSL                1455.687019
               Yes           DSL                2542.344194
                             Fiber optic        3147.553430
                             No                  932.299038
Name: TotalCharges, dtype: float64

Dapat dilihat nilai rata-rata **TOTAL CHARGES** dari customer **NON-SENIOR CITIZEN (Senior Citizen = 0)** yang memiliki **PHONE SERVICE** dan memiliki **INTERNET SERVICE FIBER OPTIC** adalah sebesar 3226.492

### Silakan menjawab di bawah ini:

#### Soal 1:
Buanglah baris dengan tenure samadengan 0 (jadi, data yang akan kita analisa di PR ini adalah baris-baris dengan tenure TIDAK samadengan 0). Ada berapa baris yang sekarang tersedia di data kita?

#### Jawaban: 7032 (langsung edit di sini aja)

#### Soal 2:
Bagi customer yang TIDAK MEMILIKI PARTNER dan TIDAK MEMILIKI DEPENDENTS, berapa rata-rata dari tenure mereka?

#### Jawaban: 23.303

#### Soal 3:
Bagi customer yang MEMILIKI PARTNER dan TIDAK MEMILIKI DEPENDENTS, berapa maksimal MonthlyCharges yang tercata?

#### Jawaban: 118.60

#### Soal 4:
Berapa banyak customer yang TIDAK MEMILIKI PARTNER namun MEMILIKI DEPENDENTS?

#### Jawaban: 359

#### Soal 5:
Berapa rata-rata TOTAL CHARGES dari customer NON-SENIOR CITIZEN (Senior Citizen = 0) yang memiliki PHONE SERVICE dan memiliki INTERNET SERVICE FIBER OPTIC?

#### Jawaban: 3226.492